src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackEncoder.Dynamic.cs

```diff
@@ -0,0 +1,289 @@
+        {
+            if (staticTableIndex != -1)
+            {
+                // Prefer static table index.
+                return staticTableIndex;
+            }
+
+            return CalculateDynamicTableIndex(name);
+        }
+
+        private bool EncodeDynamicHeader(Span<byte> buffer, int staticTableIndex, string name, string value, out int bytesWritten)
+        {
+            var headerField = GetEntry(name, value);
+            if (headerField != null)
+            {
+                // Already exists in dynamic table. Write index.
+                var index = CalculateDynamicTableIndex(headerField.Index);
+                return EncodeIndexedHeaderField(index, buffer, out bytesWritten);
+            }
+            else
+            {
+                // Doesn't exist in dynamic table. Add new entry to dynamic table.
+                var headerSize = (uint)HeaderField.GetLength(name.Length, value.Length);
+
+                var index = ResolveDynamicTableIndex(staticTableIndex, name);
+                var success = index == -1
+                    ? EncodeLiteralHeaderFieldIndexingNewName(name, value, buffer, out bytesWritten)
+                    : EncodeLiteralHeaderFieldIndexing(index, value, buffer, out bytesWritten);
+
+                if (success)
+                {
+                    EnsureCapacity(headerSize);
+                    AddHeaderEntry(name, value, headerSize);
+                }
+
+                return success;
+            }
+        }
+
+        /// <summary>
+        /// Ensure there is capacity for the new header. If there is not enough capacity then remove
+        /// existing headers until space is available.
+        /// </summary>
+        private void EnsureCapacity(uint headerSize)
+        {
+            Debug.Assert(headerSize <= _maxHeaderTableSize, "Header is bigger than dynamic table size.");
+
+            while (_maxHeaderTableSize - _headerTableSize < headerSize)
+            {
+                var removed = RemoveHeaderEntry();
+                Debug.Assert(removed != null);
+
+                // Removed entries are tracked to be reused.
+                PushRemovedEntry(removed);
+            }
+        }
+
+        private EncoderHeaderEntry? GetEntry(string name, string value)
+        {
+            if (_headerTableSize == 0)
+            {
+                return null;
+            }
+            var hash = name.GetHashCode();
+            var bucketIndex = CalculateBucketIndex(hash);
+            for (var e = _headerBuckets[bucketIndex]; e != null; e = e.Next)
+            {
+                // We've already looked up entries based on a hash of the name.
+                // Compare value before name as it is more likely to be different.
+                if (e.Hash == hash &&
+                    string.Equals(value, e.Value, StringComparison.Ordinal) &&
+                    string.Equals(name, e.Name, StringComparison.Ordinal))
```

Comment:
Ordinal is the default, and specifying it explicitly will be a bit more expensive.  These could just be `string.Equals(value, e.Value) && string.Equals(name, e.Name)`, or even `value == e.Value && name == e.Name`.